microprofile/WFLY-19954_Preview_Support_vertx_feature_pack.adoc

```diff
@@ -0,0 +1,234 @@
+
+```java
+@Any @io.smallrye.common.annotation.Identifier.Literal.of("vertx");
+```
+
+The exposed Vertx instance is mainly used by other subsystems which needs a Vertx instance, like https://github.com/wildfly/wildfly/tree/main/microprofile/telemetry-smallrye[microprofile-telemetry] and https://github.com/wildfly/wildfly/tree/main/microprofile/reactive-messaging-smallrye[microprofile-reactive-messaging], and they don't need to depend on this feature pack from the source code perspective.
+
+The idea of introducing the qualifier is to make CDI lookup side can opt in until this feature pack has high stability level.
+
+The qualifier was selected after discussions with SmallRye team on the related issues above.
+
+==== Get the Vertx instance with the capability name in other subsystems
+The Vertx instance is also a MSC Service so that other services can access it with the capability name:
+[source,java]
+----
+CapabilityServiceBuilder<?> sb = context.getCapabilityServiceTarget().addService();
+Supplier<VertxProxy> vertxProxySupplier = sb.requiresCapability("org.wildfly.extension.vertx", VertxProxy.class);
+----
+NOTE: The Vertx MSC Service is only available when it is set up.
+
+=== Nice-to-Have Requirements
+As the Vertx instance is exposed to the CDI container, it would be good if it can be accessed in the application codes. As the usage of the Vertx instance in application code is undefined, a warning log message can be printed to notify that it is experimental.
+
+Besides configuration each field in the management model, it would be good to configure the {vertx_options_with_link} from a JSON file in the server's configuration directory.
+
+=== Changed requirements
+N/A
+
+=== Non-Requirements
+There is no plan currently of a clustered Vertx instance within WildFly, so some clustering related options will not be configured using `vertx` subsystem, including `EventBusOptions`.
+
+`MetricsOptions` is not required to configure.
+`TracingOptions` is not required to configure.
+
+=== Future Work
+The initial stability level is targeting `Preview`, and the plan is to promote it to higher stability level after some baking time in community.
+
+New attributes maybe added in the future according to the changes in the {vertx_options_with_link}, the vertx subsystem needs to be updated manually to match the new fields.
+
+== Backwards Compatibility
+N/A
+
+=== Default Configuration
+The `vertx` extension/subsystem needs to be activated explicitly to work:
+[source,shell]
+----
+if (outcome != success) of /subsystem=vertx:read-resource
+  echo INFO: Adding vertx subsystem.
+  /extension=org.wildfly.extension.vertx:add
+  /subsystem=vertx:add
+  :reload
+else
+  echo INFO: vertx subsystem already in configuration, subsystem not added.
+end-if
+----
+
+There is no Vertx instance set up by default, users need to do it explicitly like:
+[source,shell]
+----
+/subsystem=vertx/vertx=vertx:add()
```

Comment:
@jasondlee @kabir Do you see any options that we need to stay using the default ones ?  

I am not sure if the following options should be updated by users or not:
* `disableTCCL` - defaults to `false`, so a context will be associated to the TCCL when it was created.
* `useDaemonThread` - defaults to `false`, so all Vertx threads are not daemons.

Also do you see some system properties that we need to set before constructing the Vertx instance in microprofile-opentelemetry and microprofile-reactive-messaging subsystem ?

One example I see coming from opentelemetry subsystem[1]:
```java
        // We need to disable vertx's DNS resolver as it causes failures under k8s
        if (System.getProperty(VERTX_DISABLE_DNS_RESOLVER) == null) {
            System.setProperty(VERTX_DISABLE_DNS_RESOLVER, "true");
        }
```

[1] https://github.com/wildfly/wildfly/blob/34.0.0.Final/observability/opentelemetry/src/main/java/org/wildfly/extension/opentelemetry/OpenTelemetrySubsystemRegistrar.java#L184

Reply:
At least for reactive messaging I'm not setting any properties in WildFly. I don't know if any are set internally in SmallRye RM. I guess not since when running in Quarkus they use the Vert.X instance provided by Quarkus

Reply:
> I am not sure if the following options should be updated by users or not:

Those sound like properties that I'm not sure we'd want users managing inside an application server, but I could be wrong.

> One example I see coming from opentelemetry subsystem[1]:

That's kind of an ugly hack that we had to put in because vertx did not (and may still not. I haven't looked) provide an API to set. I don't know if setting that property in `opentelemetry` will affect the vertx instance from the FP. I'd be surprised if it's set in time, given the module system lifecycle, etc., so we might either need to make that a configurable option, or just add that to the FP.

Reply:
Maybe we should keep the 2 options(`disableTCCL`, `useDaemonThread`) untouched by users, and open only when there are explicit demands to configure them.

Reply:
> because vertx did not (and may still not. I haven't looked) provide an API to set ...
> so we might either need to make that a configurable option, or just add that to the FP.

I don't see any API either to set, so probably be good to add that to the FP as well before a Vertx instance is constructed so that it takes effect.

Reply:
For those curious and unfamiliar with this issue, the code in question is here: https://github.com/eclipse-vertx/vert.x/blob/4.x/src/main/java/io/vertx/core/spi/resolver/ResolverProvider.java#L35-L50

Reply:
I vote for keeping those options untouched for now. We can always add later